hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java

```diff
@@ -618,23 +618,31 @@ public boolean isBucketEmpty(String volume, String bucket)
     }
     int currentCount = 0;
 
-    try (TableIterator<String, ? extends KeyValue<String, OmBucketInfo>>
-        bucketIter = bucketTable.iterator()) {
-      KeyValue<String, OmBucketInfo> kv = bucketIter.seek(startKey);
-      while (currentCount < maxNumOfBuckets && bucketIter.hasNext()) {
-        kv = bucketIter.next();
-        // Skip the Start Bucket if needed.
-        if (kv != null && skipStartKey &&
-            kv.getKey().equals(startKey)) {
+
+    // For Bucket it is full cache, so we can just iterate in-memory table
+    // cache.
+    Iterator<Map.Entry<CacheKey<String>, CacheValue<OmBucketInfo>>> iterator =
```

Comment:
Sorry, I am not about to make sure of thi; but for each request do we iterator through the whole bucket space here?

Reply:
There is a maxCount, when we reach the count, we will return immediately, in that case, we shall not iterate all volumes in a bucket.

BucketTable cache is just a concurrentHashMap of all buckets in OM.

So let's take an example: As it is ConcurrentSkipListMap, it has items sorted based on the key.

We have entries in bucket table cache like below: (For brevity, removed bucketInfo structures which are the values for the key.)
/vol/buck1
/vol/buck2
/vol/buck3
/vol2/bucket2
/vol2/bucket3
/vol2/bucket4

When we want to list buckets of /vol, and in each return only 1 entry(maximum count), we return /vol/buck1 and then immediately return, next time listBuckets is called, startkey will be /vol/buck1,1, we return /vol/buck2, (To return this we iterate 2 entries), like that so on.

But when we want to list all buckets is /vol2, we will iterate the entries from start, and reach to /vol2 in the cache and once the maximum count is reached we return from there.

So, to answer your question not every time we iterate entire cache map.

Reply:
To improve this API, I think we should change our cache data structure which will be effective for both read/write/list API. (Initially, usage of ConcurrentHashMap, as get() is a constant time operation, but it has caused the problem for list API).

Reply:
> But when we want to list all buckets is /vol2, we will iterate the entries from start, and reach to /vol2 in the cache and once the maximum count is reached we return from there.

The architecture of SkipList prevents us from iterating all the keys. That is good enough, I was worried that we will walk all the entries. I missed we were using a skipList based hash table.

Reply:
You can consider this comment resolved. Thanks for the explanation. I am leaving it open for other reviewers who might want to read this patch.